src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,224 @@
+#include <cstdint>
+#include <memory>
+#include <span>
+#include <system_error>
+#include <utility>
+
+#include <spdlog/spdlog.h>
+
+#include "ClpJsException.hpp"
+#include "constants.hpp"
+
+using namespace std::literals::string_literals;
+
+auto ClpIrV1Decoder::create(emscripten::val const& data_array) -> ClpIrV1Decoder {
+    auto const length{data_array["length"].as<size_t>()};
+    SPDLOG_INFO("ClpIrV1Decoder::ClpIrV1Decoder() got buffer of length={}", length);
+
+    auto data_buffer{std::make_unique<char const[]>(length)};
+    emscripten::val::module_property("HEAPU8")
+            .call<void>("set", data_array, reinterpret_cast<uintptr_t>(data_buffer.get()));
+
+    auto const zstd_decompressor{std::make_shared<clp::streaming_compression::zstd::Decompressor>()
+    };
+    zstd_decompressor->open(data_buffer.get(), length);
+
+    bool is_four_bytes_encoding{true};
+    if (auto const err{
+                clp::ffi::ir_stream::get_encoding_type(*zstd_decompressor, is_four_bytes_encoding)
+        };
+        clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success != err)
+    {
+        SPDLOG_CRITICAL("Failed to decode encoding type, err={}", err);
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decode encoding type.."
+        );
+    }
+
+    if (false == is_four_bytes_encoding) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Is not four byte encoding."
+        );
+    }
+    auto result{clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>::create(
+            *zstd_decompressor
+    )};
+    if (result.has_error()) {
+        auto const error_code{result.error()};
+        SPDLOG_CRITICAL(
+                "Failed to decompress: {}:{}",
+                error_code.category().name(),
+                error_code.message()
+        );
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decompress"
+        );
+    }
+
+    return std::move(
+            ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()))
+    );
+}
+
+auto ClpIrV1Decoder::get_estimated_num_events() const -> size_t {
+    return m_log_events.size();
+}
+
+auto ClpIrV1Decoder::build_idx(size_t begin_idx, size_t end_idx) -> emscripten::val {
+    constexpr size_t cFullRangeEndIdx{0};
+    if (0 != begin_idx && cFullRangeEndIdx != end_idx) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Partial range index building is not yet supported."
+        );
+    }
+    if (false == m_full_range_built) {
+        m_full_range_built = true;
```

Comment:
Maybe move this line to the end of the `if` when everything successes?

Reply:
I was thinking that we want users to avoid calling `build_idx` multiple times on the full range, before we add partial range support. If any deserialization fails and the flag does not get set, `build_idx` can still be invoked. (Note that `m_full_range_built` is really a temporary solution; we will add support for partial range building eventually, when we start to optimize. )